src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -80,7 +107,53 @@ public override void Dispose()
 
         protected override bool OnSysCall(uint method)
         {
-            return InteropService.Invoke(this, method);
+            if (!services.TryGetValue(method, out InteropDescriptor descriptor))
+                return false;
+            if (!descriptor.AllowedTriggers.HasFlag(Trigger))
+                return false;
+            ExecutionContextState state = CurrentContext.GetState<ExecutionContextState>();
+            if (!state.CallFlags.HasFlag(descriptor.RequiredCallFlags))
+                return false;
+            if (!AddGas(descriptor.FixedPrice))
+                return false;
+            List<object> parameters = descriptor.Parameters.Length > 0
+                ? new List<object>()
+                : null;
+            foreach (var pd in descriptor.Parameters)
+            {
+                StackItem item = Pop();
+                object value;
+                if (pd.IsArray)
+                {
+                    Array av;
+                    if (item is VMArray array)
+                    {
+                        av = Array.CreateInstance(pd.Type.GetElementType(), array.Count);
+                        for (int i = 0; i < av.Length; i++)
+                            av.SetValue(pd.Converter(array[i]), i);
+                    }
+                    else
+                    {
+                        av = Array.CreateInstance(pd.Type.GetElementType(), (int)item.GetBigInteger());
+                        for (int i = 0; i < av.Length; i++)
+                            av.SetValue(pd.Converter(Pop()), i);
+                    }
+                    value = av;
+                }
+                else
+                {
+                    value = pd.Converter(item);
+                    if (pd.IsEnum)
+                        value = Convert.ChangeType(value, pd.Type);
+                    else if (pd.IsInterface)
+                        value = ((InteropInterface)value).GetInterface<object>();
+                }
+                parameters.Add(value);
+            }
+            object returnValue = descriptor.Handler.Invoke(this, parameters?.ToArray());
```

Comment:
I think that this dynamic calls could be slower

Reply:
```c#
class TestClass
{
    private int Method(int a) => a;

    public void Test()
    {
        MethodInfo method = GetType().GetMethod(nameof(Method), BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
        Stopwatch stopwatch = Stopwatch.StartNew();
        for (int i = 0; i < 1_000_000; i++)
        {
            method.Invoke(this, new object[] { i });
        }
        stopwatch.Stop();
        Console.WriteLine($"{stopwatch.ElapsedMilliseconds}ms"); //169ms
    }
}
```

One million calls take only 169 ms, which is much higher than our target tps, so maybe we don't need to worry about the performance loss here first.